llvm/lib/IR/Verifier.cpp

```diff
@@ -2086,6 +2088,27 @@ void Verifier::checkUnsignedBaseTenFuncAttr(AttributeList Attrs, StringRef Attr,
   }
 }
 
+void Verifier::verifyX86ABI(FunctionType *FT, AttributeList Attrs,
+                            const Value *V, unsigned MaxParameterWidth) {
+  if (!Attrs.hasFnAttr("target-features"))
+    return;
+
+  StringRef TF = Attrs.getFnAttr("target-features").getValueAsString();
+  // Check SSE feature.
+  Check(!TT.isArch64Bit() || !TF.contains("-sse,") ||
+        !FT->getReturnType()->isFloatTy(),
+        "SSE register return with SSE disabled", V);
+  // Check SSE2 feature.
+  Check(!TT.isArch64Bit() || !TF.contains("-sse2") ||
+        (!FT->getReturnType()->isDoubleTy() &&
+         !FT->getReturnType()->is16bitFPTy()),
+        "SSE2 register return with SSE2 disabled", V);
+  // Check EVEX512 feature.
+  if (MaxParameterWidth >= 512)
+    Check(!TF.contains("+avx512f") || !TF.contains("-evex512"),
+          "512-bit vector arguments require 'evex512' for AVX512", V);
+}
```

Comment:
I don't think anything like this belongs in the IR verifier. It should not be invalid IR. You can always codegen it to something (whether or not some platform ABI defined how to do this is entirely irrelevant)

Reply:
https://github.com/llvm/llvm-project/issues/111406 is a good illustration to this question. Arbitrarily generating something is unexpected here.

Reply:
This is clang backendification. No type should require any specific target properties.

This isn't really a valid way to check for subtarget features either. You have to parse out the whole string and evaluate the final result. e.g. you can do -feature,+feature and +feature wins

Reply:
> This is clang backendification. No type should require any specific target properties.

Some type does require specific target properties, e.g., FP16/BF16, which we have limited them in Clang front end.

> This isn't really a valid way to check for subtarget features either. You have to parse out the whole string and evaluate the final result. e.g. you can do -feature,+feature and +feature wins

Good catch, done.